sound/soc/sof/intel/hsw.c

```diff
@@ -93,8 +93,13 @@ static void hsw_block_write(struct snd_sof_dev *sdev, u32 offset, void *src,
 	__iowrite32_copy(dest, src, m);
 
 	if (n) {
-		for (i = 0; i < n; i++)
-			tmp |= (u32)*(src_byte + m * 4 + i) << (i * 8);
+		/* first read the 32bit data of dest, then change affected
+		 * bytes, and write back to dest. For unaffected bytes, it
+		 * should not be changed
+		 */
+		__ioread32_copy(&tmp, dest + m * 4, 1);
+
+		tmp |= *(u32 *)(src_byte + m * 4) & ((1 << (8 * n)) - 1);
```

Comment:
@keyonjie what you have here is correct and i've tested it to work.
But I have a suggestion to improve the readability. 

tmp |= *(u32 *)(src_byte + m * 4) & GENMASK(31, 32 - n * 8 );

Reply:
@ranj063 is the GENMASK arguments correct? For n=1 this should be 1<<8-1, so 0xFF

Reply:
@ranj063 @plbossart the mask is start from bit 0, so for n > 0, it should be GENMASK((8 * n) - 1, 0). But we can't use GENMASK if the mask is 0.

Reply:
oops looks like I misunderstood it. Yes but maybe like @plbossart said it is better to leave the array as it is definitely easier to follow.